### PR TITLE
Documentation for the astropy modeling JointFitter

### LINKED ISSUE #3445: Remove JointFitter
What the `JointFitter` does is equivalent to fitting a compound joined `&` model. When The `JointFitter` class should be removed when compound joined models become fittable. This is also related to #3444.

### PR DESCRIPTION
**Pull request by karllark**
### Description

This pull request provides documentation for the existing astropy.modeling JointFitter.

Close #3445

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: docs/modeling/jointfitter.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+   plt.figure(figsize=(8,5))
+   plt.plot(x1, y1, 'bo', alpha=0.25)
+   plt.plot(x2, y2, 'go', alpha=0.25)
+   plt.plot(x1, gjf1(x1), 'b--', label='AreaGaussian1')
+   plt.plot(x2, gjf2(x2), 'g--', label='AreaGaussian2')
+   plt.xlabel('Wavelength')
+   plt.ylabel('Flux')
+   plt.legend(loc=2)
+
+.. image:: example_spectal_segments_jointfitter.png
```
Comment (pllim) [inline docs/modeling/jointfitter.rst:189]: Why provide a static image when you can have matplotlib sphinx plot directive to do it for you?

**Review by pllim — COMMENTED** [review]
Is there a way to write this RST in a way that we can doctest?

**Comment by karllark** [comment]
> Is there a way to write this RST in a way that we can doctest?

Not sure.  Suggestions welcome!

**Comment by karllark** [comment]
> > Is there a way to write this RST in a way that we can doctest?
> 
> Not sure. Suggestions welcome!

If AreaGaussian1D didn't need to be defined, then the `>>>` method could work.

One idea would be to add AreaGaussian1D (possible with a better name) to the astropy models as part of this PR.  Not sure that is the right way to go, though.

**Comment by pllim** [comment]
Ah, to combine matplotlib *and* doctestplus is tricky. I don't think they play well with each other. In the past I have repeated the code twice, once with `>>>` in narrative docs for those I want to doctest, and again in `.. plot:` for those I want to use for the plot.

**Comment by pllim** [comment]
@nstarman, @adrn, et al. have done a lot of improvements to docs/Learn recently, maybe they have ideas?

**Comment by nstarman** [comment]
If I recall correctly, part of the issue is that code-blocks are executed under different conditions. With sphinx extensions they can be executed, but with astropy's doctest they are not executed unless prepended by ``>>>``.
You could try seeing if an RTD build will generate the image given code in ``.. code-block::``, even if that code isn't "tested".

File: docs/modeling/jointfitter.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+The resulting fit parameters show that the area and mean wavelength of the
+two AreaGaussian1D models are exactly the same while the width (stddev) is
+different reflecting the different spectral resolutions of the two segments.
+
+.. code-block:: python
+
+   print("AreaGaussian1 parameters")
+   print(gjf1.param_names)
+   print(gjf1.parameters)
+   print("AreaGaussian1 parameters")
```
Comment (nden) [inline docs/modeling/jointfitter.rst:161]: Did you mean `AreaGaussian2`?

File: docs/modeling/jointfitter.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+datasets.
+
+.. code-block:: python
+
+   # define the two models to be fit
+   gjf1 = AreaGaussian1D(
+       area=1.0, mean=5.0, stddev=1.0, bounds={"area": (0, 10), "stddev": (0.1, 0.3)}
+   )
+   gjf2 = AreaGaussian1D(
+       area=1.0, mean=5.0, stddev=1.0, bounds={"area": (0, 10), "stddev": (0.1, 0.5)}
```
Comment (nden) [inline docs/modeling/jointfitter.rst:144]: I forget, do bounds on individual models have an effect when used with the JointFitter?

File: docs/modeling/jointfitter.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+   print("AreaGaussian1 parameters")
+   print(gjf1.param_names)
+   print(gjf2.parameters)
+
+::
+
+   AreaGaussian1 parameters
+   ('area', 'mean', 'stddev')
+   [1.48697226 5.09826068 0.19761087]
+   AreaGaussian1 parameters
```
Comment (nden) [inline docs/modeling/jointfitter.rst:170]: `AreaGaussian2`

**Review by nden — COMMENTED** [review]
A nice read, thanks!
Just a minor comment to address.

File: docs/modeling/jointfitter.rst (hunk 2 above)
Comment (karllark) [reply docs/modeling/jointfitter.rst:161]: Yep.  Fixed.

File: docs/modeling/jointfitter.rst (hunk 3 above)
Comment (karllark) [reply docs/modeling/jointfitter.rst:144]: No effect.  Checked and added the note near the beginning of this file to clearly state this.

Removed bounds definition to make this example cleaner.

File: docs/modeling/jointfitter.rst (hunk 4 above)
Comment (karllark) [reply docs/modeling/jointfitter.rst:170]: Fixed.

**Review by nden — APPROVED** [review]

**Comment by pllim** [comment]
None of this example appears to be doctested. If it breaks, we'll never know until someone actually runs it. FYI.

**Pull request merged**